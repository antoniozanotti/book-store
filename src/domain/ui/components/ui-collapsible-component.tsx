"use client";

import { ChevronDownIcon, ChevronLeft } from "lucide-react";
import { useState, ReactNode, useRef } from "react";

export interface UiCollapsibleComponentInterface
  extends React.ComponentProps<"div"> {
  shouldCloseWhenOutside: boolean;
  theHeader: ReactNode;
  theContent: ReactNode;
}

export function UiCollapsibleComponent({
  shouldCloseWhenOutside,
  theHeader,
  theContent,
  ...props
}: UiCollapsibleComponentInterface) {
  const [isOpen, setIsOpen] = useState(false);
  const timeoutRef = useRef<number | null>(null);

  function onClickHandler() {
    if (isOpen) {
      setIsOpen(false);
    } else {
      setIsOpen(true);
    }
  }

  function onFocusHandler() {
    if (timeoutRef.current !== null) {
      clearTimeout(timeoutRef.current);
    }
  }

  function onBlurHandler() {
    timeoutRef.current = window.setTimeout(() => {
      setIsOpen(false);
    });
  }

  const contentClasses = isOpen ? "lg:absolute lg:z-[100]" : "hidden";

  return (
    <div {...props} onBlur={onBlurHandler} onFocus={onFocusHandler}>
      <button
        className="flex cursor-pointer gap-2 p-4 lg:px-10 lg:py-2 items-center"
        onClick={onClickHandler}
        tabIndex={0}
        aria-haspopup="true"
        aria-expanded={isOpen}
      >
        <span className="grow">{theHeader}</span>
        {isOpen ? (
          <ChevronDownIcon className="size-4" />
        ) : (
          <ChevronLeft className="size-4" />
        )}
      </button>
      <div className={contentClasses}>{theContent}</div>
    </div>
  );
}
