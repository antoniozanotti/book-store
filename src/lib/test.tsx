import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
