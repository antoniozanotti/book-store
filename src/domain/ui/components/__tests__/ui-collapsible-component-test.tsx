import { render, screen } from "@/lib/test";
import { UiCollapsibleComponent } from "../ui-collapsible-component";

describe("UiCollapsibleComponent", () => {
  it("should render", async () => {
    const { container } = render(<UiCollapsibleComponent theContent="Content" theHeader="Header" />);
    expect(container).toMatchSnapshot();
  });
});