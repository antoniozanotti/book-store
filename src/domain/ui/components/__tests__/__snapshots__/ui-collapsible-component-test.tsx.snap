// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiCollapsibleComponent should render 1`] = `
<div>
  <div
    data-testid="container"
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="flex cursor-pointer gap-2 p-4 lg:px-10 lg:py-2 items-center"
        tabindex="0"
      >
        <span
          class="grow"
        >
          Header
        </span>
        <svg
          class="lucide lucide-chevron-left size-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m15 18-6-6 6-6"
          />
        </svg>
      </button>
      <div
        class="hidden"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;
